frontend/controller/actions/group.js

```diff
@@ -0,0 +1,89 @@
+import {
+  INVITE_INITIAL_CREATOR,
+  createInvite
+} from '@model/contracts/group.js'
+import proposals, {
+  PROPOSAL_INVITE_MEMBER,
+  PROPOSAL_REMOVE_MEMBER,
+  PROPOSAL_GROUP_SETTING_CHANGE,
+  PROPOSAL_PROPOSAL_SETTING_CHANGE,
+  PROPOSAL_GENERIC
+} from '@model/contracts/voting/proposals.js'
+import { RULE_THRESHOLD } from '@model/contracts/voting/rules.js'
+import { GIErrorUIRuntimeError } from '@model/errors.js'
+
+import imageUpload from '@utils/imageUpload.js'
+import { merge } from '@utils/giLodash.js'
+// import { CONTRACT_IS_SYNCING } from '@utils/events.js'
+import L from '@view-utils/translations.js'
+
+export default sbp('sbp/selectors/register', {
+  'gi.actions/group/create': async function ({
+    name,
+    picture,
+    sharedValues,
+    mincomeAmount,
+    mincomeCurrency,
+    thresholdChange,
+    thresholdMemberApproval,
+    thresholdMemberRemoval
+  }, cb) {
+    let finalPicture = `${window.location.origin}/assets/images/default-group-avatar.png`
+
+    if (picture) {
+      try {
+        finalPicture = await imageUpload(picture)
+      } catch (e) {
+        console.error('Failed to upload the group picture', e)
+        throw new GIErrorUIRuntimeError(L('Failed to upload the group picture. {codeError}', { codeError: e.message }))
+      }
+    }
+
+    try {
+      const initialInvite = createInvite({ quantity: 60, creator: INVITE_INITIAL_CREATOR })
+      const entry = sbp('gi.contracts/group/create', {
+        invites: {
+          [initialInvite.inviteSecret]: initialInvite
+        },
+        settings: {
+          // authorizations: [contracts.CanModifyAuths.dummyAuth()], // TODO: this
+          groupName: name,
+          groupPicture: finalPicture,
+          sharedValues,
+          mincomeAmount: +mincomeAmount, // ensure this is a number
+          mincomeCurrency: mincomeCurrency,
+          proposals: {
+            // TODO: make the UI support changing the rule type, so that we have
+            //       a component for RULE_DISAGREEMENT as well
+            [PROPOSAL_GROUP_SETTING_CHANGE]: merge({},
+              proposals[PROPOSAL_GROUP_SETTING_CHANGE].defaults,
+              { ruleSettings: { [RULE_THRESHOLD]: { threshold: thresholdChange } } }
+            ),
+            [PROPOSAL_INVITE_MEMBER]: merge({},
+              proposals[PROPOSAL_INVITE_MEMBER].defaults,
+              { ruleSettings: { [RULE_THRESHOLD]: { threshold: thresholdMemberApproval } } }
+            ),
+            [PROPOSAL_REMOVE_MEMBER]: merge({},
+              proposals[PROPOSAL_REMOVE_MEMBER].defaults,
+              { ruleSettings: { [RULE_THRESHOLD]: { threshold: thresholdMemberRemoval } } }
+            ),
+            [PROPOSAL_PROPOSAL_SETTING_CHANGE]: proposals[PROPOSAL_PROPOSAL_SETTING_CHANGE].defaults,
+            [PROPOSAL_GENERIC]: proposals[PROPOSAL_GENERIC].defaults
+          }
+        }
+      })
+      await sbp('backend/publishLogEntry', entry)
+      return entry.hash()
```

Comment:
@taoeffect please review this part, is this what you suggested?
I don't like that much this approach because when creating a group, it's part of the "core logic" to switch to it. So, I think that part and the `contract/subscribeAndWait` should be part of it.
I understand why we did it - because of the tricky part with the callback - but still... we should look for a cleaner solution where the logic on the Vue components is as minimal as possible.

Reply:
Yes, this is what I had in mind. It's *critically* important that *actions do only what their names say they will do*. We should add that somewhere to our style guidelines. This will prevent programmer confusion resulting form unexpected behavior. Since the action is named `create`, it should only create the group, and do nothing else.

If you want to keep the logic out of the Vue components, that is totally fine, but the way to do that is to similarly create a differently named action that does what you want it to do.

So something like:

```js
  'gi.actions/group/createAndSyncAndSwitch': async function (groupData: Object) {
    const groupID = await sbp('gi.actions/group/create', groupData)
    // assuming 'gi.actions/contract/sync' and 'gi.actions/contract/syncAndWait'
    // do the same thing, we can just call 'gi.actions/contract/sync',
    // otherwise we would need to call 'gi.actions/contract/syncAndWait' here
    await sbp('gi.actions/contract/sync', groupID)
    sbp('gi.actions/group/switch', groupID)
  }
```

Note that if you implement `'gi.actions/group/createAndSync'`, then this selector should be implemented using that, so we have something like this instead:

```js
  'gi.actions/group/createAndSync': async function (groupData: Object) {
    const groupID = await sbp('gi.actions/group/create', groupData)
    // assuming 'gi.actions/contract/sync' and 'gi.actions/contract/syncAndWait'
    // do the same thing, we can just call 'gi.actions/contract/sync',
    // otherwise we would need to call 'gi.actions/contract/syncAndWait' here
    return sbp('gi.actions/contract/sync', groupID)
  },
  'gi.actions/group/createAndSyncAndSwitch': async function (groupData: Object) {
    const groupID = await sbp('gi.actions/group/createAndSync', groupData)
    sbp('gi.actions/group/switch', groupID)
  }
```

Note that here we assume that `'gi.actions/contract/sync'` does the same thing as the `'gi.actions/contract/syncAndWait'` implementation I gave in the other comment. You'll need to verify that it does first. If it doesn't, you'll need to replace it here with `'gi.actions/contract/syncAndWait'`.

Finally, note that doing it like this also means that the contract should be fully synced when we return from either `'gi.actions/group/createAndSync'` or `'gi.actions/group/createAndSyncAndSwitch'`, which means that the code you have below in `CreateGroup.vue` should be rewritten, i.e.:

```js
const groupID = await sbp('gi.actions/group/create', { ... })
sbp('okTurtles.events/once', groupID, (contractID, entry) => {
  sbp('gi.actions/group/switch', groupID)
  this.next()
})
await sbp('gi.actions/contract/subscribeAndWait', groupID)

// BECOMES:
const groupID = await sbp('gi.actions/group/createAndSyncAndSwitch', { ... })
this.next()

// OR:
const groupID = await sbp('gi.actions/group/createAndSync', { ... })
sbp('gi.actions/group/switch', groupID)
this.next()
```